File: analysis-service/app/services/explanation_service.py
from typing import Any

from openai import AsyncOpenAI

from app.config import get_settings
from app.services.embedding_service import get_embedding_service
from app.services.faiss_store import FAISSStore
from app.utils.log_formatter import format_activity_log, format_activity_logs


_SYSTEM_PROMPT = """\
You are a cybersecurity analyst assistant embedded in SentinelX, a Zero-Trust \
Identity Intelligence Platform. Your sole role is to explain why a specific \
security event looks suspicious by comparing it against the user's historical \
normal behaviour.

Rules:
- Be concise (2–4 sentences maximum).
- Focus on concrete deviations: unusual time, unfamiliar endpoint, new IP, \
abnormal frequency.
- Do NOT make security decisions, assign blame, or recommend actions.
- Do NOT declare the user guilty or innocent.
- Write in plain English for a security operations dashboard.\
"""


def _build_prompt(current_text: str, history_texts: list[str]) -> str:
    history_block = "\n".join(f"  • {t}" for t in history_texts)
    return (
        f"Normal historical behaviour for this user:\n{history_block}\n\n"
        f"Current suspicious event:\n  {current_text}\n\n"
        "Explain why this event deviates from the user's normal behaviour."
    )


class ExplanationService:
    """
    Stateless RAG pipeline:
      1. Embed the suspicious event.
      2. Build a per-request FAISS index from historical (normal) logs.
      3. Retrieve the top-k most similar historical logs.
      4. Ask OpenAI to explain the deviation in plain English.
    """

    def __init__(self) -> None:
        self._settings = get_settings()
        self._embedding_service = get_embedding_service()
        self._openai: AsyncOpenAI | None = None

    @property
    def _client(self) -> AsyncOpenAI:
        if self._openai is None:
            self._openai = AsyncOpenAI(api_key=self._settings.openai_api_key)
        return self._openai

    async def explain(
        self,
        current_event: dict[str, Any],
        historical_logs: list[dict[str, Any]],
    ) -> dict[str, Any]:
        if not historical_logs:
            return {
                "explanation": (
                    "No historical behaviour data is available for this user. "
                    "Unable to perform behavioural comparison."
                ),
                "retrieved_logs": [],
                "current_event_text": format_activity_log(current_event),
            }

        # --- Embed historical logs and build a fresh FAISS index ---
        historical_texts = format_activity_logs(historical_logs)
        historical_embeddings = self._embedding_service.embed_texts(historical_texts)

        dimension = historical_embeddings.shape[1]
        store = FAISSStore(dimension)
        store.add_logs(historical_logs, historical_embeddings)

        # --- Embed current event and retrieve nearest historical logs ---
        current_text = format_activity_log(current_event)
        current_embedding = self._embedding_service.embed_text(current_text)
        retrieved = store.search(current_embedding, k=self._settings.faiss_top_k)
        retrieved_texts = format_activity_logs(retrieved)

        # --- Build prompt and call OpenAI ---
        prompt = _build_prompt(current_text, retrieved_texts)

        response = await self._client.chat.completions.create(
            model=self._settings.openai_model,
            messages=[
                {"role": "system", "content": _SYSTEM_PROMPT},
                {"role": "user", "content": prompt},
            ],
            temperature=0.3,
            max_tokens=220,
        )

        explanation = response.choices[0].message.content.strip()

        return {
            "explanation": explanation,
            "retrieved_logs": retrieved,
            "current_event_text": current_text,
        }

File: backend/src/services/explanationService.js
const OpenAI = require("openai");
const env = require("../config/env");
const { formatActivityLog, formatActivityLogs } = require("../utils/logFormatter");

const SYSTEM_PROMPT = `You are a cybersecurity analyst assistant embedded in SentinelX, a Zero-Trust \
Identity Intelligence Platform. Your sole role is to explain why a specific \
security event looks suspicious by comparing it against the user's historical \
normal behaviour.

Rules:
- Be concise (2–4 sentences maximum).
- Focus on concrete deviations: unusual time, unfamiliar endpoint, new IP, \
abnormal frequency.
- Do NOT make security decisions, assign blame, or recommend actions.
- Do NOT declare the user guilty or innocent.
- Write in plain English for a security operations dashboard.`;

let client;
function getClient() {
  if (!client) {
    client = new OpenAI({ apiKey: env.openaiApiKey });
  }
  return client;
}

function buildPrompt(currentText, historyTexts) {
  const historyBlock = historyTexts.map((t) => `  • ${t}`).join("\n");
  return (
    `Normal historical behaviour for this user:\n${historyBlock}\n\n` +
    `Current suspicious event:\n  ${currentText}\n\n` +
    "Explain why this event deviates from the user's normal behaviour."
  );
}

function cosineSimilarity(a, b) {
  let dot = 0;
  let normA = 0;
  let normB = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }
  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}

/**
 * Ranks historicalLogs by embedding similarity to the current event and
 * returns the top-k most similar (descending). Equivalent to a per-request
 * FAISS IndexFlatIP search over L2-normalised embeddings.
 */
function topKSimilar(currentEmbedding, historicalEmbeddings, historicalLogs, k) {
  const scored = historicalLogs.map((log, i) => ({
    log,
    score: cosineSimilarity(currentEmbedding, historicalEmbeddings[i]),
  }));
  scored.sort((a, b) => b.score - a.score);
  return scored.slice(0, k).map((s) => s.log);
}

async function embedTexts(texts) {
  if (texts.length === 0) return [];
  const response = await getClient().embeddings.create({
    model: env.embeddingModel,
    input: texts,
  });
  return response.data.map((d) => d.embedding);
}

/**
 * Stateless RAG pipeline:
 *   1. Embed the suspicious event and the user's historical (normal) logs.
 *   2. Retrieve the top-k most similar historical logs by cosine similarity.
 *   3. Ask OpenAI to explain the deviation in plain English.
 *
 * @param {object} params
 * @param {number}   params.userId
 * @param {object}   params.currentEvent  - the anomalous activity_log row
 * @param {object[]} params.historicalLogs - recent normal logs for this user
 * @returns {Promise<string>} human-readable explanation
 */
async function explain({ currentEvent, historicalLogs }) {
  if (!historicalLogs || historicalLogs.length === 0) {
    return (
      "No historical behaviour data is available for this user. " +
      "Unable to perform behavioural comparison."
    );
  }

  const currentText = formatActivityLog(currentEvent);
  const historyTexts = formatActivityLogs(historicalLogs);

  const [currentEmbedding, historicalEmbeddings] = await Promise.all([
    embedTexts([currentText]).then((e) => e[0]),
    embedTexts(historyTexts),
  ]);

  const retrieved = topKSimilar(
    currentEmbedding,
    historicalEmbeddings,
    historicalLogs,
    env.ragTopK
  );
  const retrievedTexts = formatActivityLogs(retrieved);

  const prompt = buildPrompt(currentText, retrievedTexts);

  const response = await getClient().chat.completions.create({
    model: env.openaiModel,
    messages: [
      { role: "system", content: SYSTEM_PROMPT },
      { role: "user", content: prompt },
    ],
    temperature: 0.3,
    max_tokens: 220,
  });

  return response.choices[0].message.content.trim();
}

module.exports = { explain, cosineSimilarity, topKSimilar };
